vllm/model_executor/layers/quantization/bitsandbytes.py

```diff
@@ -0,0 +1,175 @@
+                       output_size: int, params_dtype: torch.dtype,
+                       **extra_weight_attrs):
+        quant_ratio = 0
+        if params_dtype.is_floating_point:
+            quant_ratio = torch.finfo(params_dtype).bits // torch.iinfo(
+                torch.uint8).bits
+        else:
+            quant_ratio = torch.iinfo(params_dtype).bits // torch.iinfo(
+                torch.uint8).bits
+
+        if input_size_per_partition * sum(
+                output_partition_sizes) % quant_ratio != 0:
+            raise ValueError(
+                "The input size is not aligned with the quantized "
+                "weight shape. ")
+        qweight = Parameter(
+            torch.empty(
+                input_size_per_partition * sum(output_partition_sizes) //
+                quant_ratio,
+                1,
+                dtype=torch.uint8,
+            ),
+            requires_grad=False,
+        )
+
+        set_weight_attrs(
+            qweight,
+            {
+                "input_dim": 0,
+                # different from other quantization methods, a tensor of shape [n,m]
+                # is qunatized to [n*m, 1] in BitsAndBytes,
+                # so the output_dim is 0
+                "output_dim": 0,
+                "bnb_quant_state": None,
+                "pack_factor": quant_ratio,
+                "use_bitsandbytes": True,
+            })
+        layer.register_parameter("qweight", qweight)
+        set_weight_attrs(qweight, extra_weight_attrs)
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+
+        orginal_type = x.dtype
+        bf_x = x.to(torch.bfloat16)
+
+        qweight = layer.qweight
+
+        quant_states = getattr(qweight, "bnb_quant_state", None)
+        if quant_states is None:
+            raise RuntimeError("Quantization state is not found.")
+
+        # check quant_states is dictionary of integer to QuantState
+        if not isinstance(quant_states, dict):
+            raise RuntimeError("Quantization state is not a dictionary.")
+        if not all(isinstance(k, int) for k in quant_states.keys()):
+            raise RuntimeError("Quantization state keys are not integers.")
+        if not all(isinstance(v, QuantState) for v in quant_states.values()):
+            raise RuntimeError("Quantization state values are not QuantState.")
+
+        pack_ratio = getattr(qweight, "pack_factor", None)
+        if pack_ratio is None:
+            raise RuntimeError("Pack ratio is not found.")
+        if not isinstance(pack_ratio, int):
+            raise RuntimeError("Pack ratio is not an integer.")
+
+        num_elements = [0] * len(quant_states)
+        for seq, quant_state in enumerate(quant_states.items()):
+            num_elements[seq] = math.prod(quant_state[1].shape) // pack_ratio
+
+        offsets = np.concatenate(([0], np.cumsum(num_elements)))
```

Comment:
this will be inefficient. We should do the following to optimize:
1. move all the checks to `create_weights`/`__init__`, we cannot have them on the hot path
2. as far as I can see, `offsets` will be a per-layer constant. pre-calculate it in `create_weights`, set it on `extra_weight_attrs` and reuse multiple times

Reply:
fixed.